File: q90-q99/q98/q98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def helper(node, lower = float('-inf'), upper = float('inf')) -> bool:
            if not node:
                return True
            val = node.val
            if val <= lower or val >= upper:
                return False
            return (helper(node.left, lower, val) and helper(node.right, val, upper))

        return helper(root)

File: q1-q99/q90-q99/q98/q98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        stack<TreeNode*> node_stack;
        TreeNode* cur = root;
        long long prev_value = (long long)INT_MIN - 1;
        while (cur != NULL || !node_stack.empty()) {
            while (cur != NULL) {
                node_stack.push(cur);
                cur = cur->left;
            }
            cur = node_stack.top();
            node_stack.pop();
            if (prev_value >= cur->val) {
                return false;
            }
            prev_value = cur->val;
            cur = cur->right;
        }
        return true;
    }
};
